[Data][Flaky] `test_simple_self_join` fails due to non-deterministic ordering

## Test Name
`test_simple_self_join[_right-_left]`

## Test Location
`python/ray/data/tests/test_join.py:281`

## Issue Description
The test fails intermittently because it compares a sorted Ray Data result against an unsorted Pandas reference result. The Pandas reference inherits non-deterministic ordering from the Ray Dataset conversion.

## Root Cause
The test has an ordering mismatch in its comparison logic:

1. `doubles.to_pandas()` (line 245) converts the Ray Dataset to a Pandas DataFrame, preserving whatever row order the distributed tasks produced (e.g., IDs 0-24, 50-99, 25-49)
2. The Pandas reference join (lines 273-279) uses `doubles_pd`, which preserves this non-deterministic row order in `expected_pd`
3. The Ray Data result `joined_pd` is sorted by "id" (line 270), producing sequential order [0-99]
4. `expected_pd` is never sorted, so it retains the jumbled order from `doubles_pd`
5. `pd.testing.assert_frame_equal(expected_pd, joined_pd_sorted)` fails because one is sorted and one isn't

From the error message:
- `expected_pd` (left): `[0-24, 50-99, 25-49]` - unsorted, preserves task completion order
- `joined_pd_sorted` (right): `[0-99]` - properly sorted

## Example Failure
```
AssertionError: DataFrame.iloc[:, 0] (column name="id") are different

DataFrame.iloc[:, 0] (column name="id") values are different (75.0 %)
[left]:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 50, 51, ..., 99, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49]
[right]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, ..., 99]
```

## Proposed Fix
Use the `rows_same` utility function from `python/ray/data/_internal/util.py:1669` which compares DataFrames ignoring row order and indices. This is the recommended approach for testing Ray Data since its interface doesn't guarantee row order.

Alternatively, sort `expected_pd` by "id" as well before comparing.

## Additional Context
This issue occurs because only one side of the comparison is sorted. The test needs to either sort both sides or use an order-independent comparison method like `rows_same`.

--- a/python/ray/data/tests/test_join.py
+++ b/python/ray/data/tests/test_join.py
@@ -264,10 +264,7 @@ def test_simple_self_join(ray_start_regular_shared_2_cpus, left_suffix, right_su
         assert 'Field "double" exists 2 times' in str(exc_info.value.cause)
     else:
 
-        joined_pd = pd.DataFrame(joined.take_all())
-
-        # Sort resulting frame and reset index (to be able to compare with expected one)
-        joined_pd_sorted = joined_pd.sort_values(by=["id"]).reset_index(drop=True)
+        joined_pd = pd.DataFrame(joined.take_all()).reset_index(drop=True)
 
         # Join using Pandas (to assert against)
         expected_pd = doubles_pd.join(
@@ -278,7 +275,7 @@ def test_simple_self_join(ray_start_regular_shared_2_cpus, left_suffix, right_su
             rsuffix=right_suffix,
         ).reset_index(drop=True)
 
-        assert rows_same(expected_pd, joined_pd_sorted), "Expected contents to be same"
+        assert rows_same(expected_pd, joined_pd), "Expected contents to be same"
 
 
 def test_invalid_join_config(ray_start_regular_shared_2_cpus):

--- a/python/ray/data/tests/test_join.py
+++ b/python/ray/data/tests/test_join.py
@@ -8,7 +8,7 @@ from packaging.version import parse as parse_version
 import ray
 from ray._private.arrow_utils import get_pyarrow_version
 from ray.data._internal.logical.operators.join_operator import JoinType
-from ray.data._internal.util import MiB
+from ray.data._internal.util import MiB, rows_same
 from ray.data.context import DataContext
 from ray.data.dataset import Dataset
 from ray.exceptions import RayTaskError
@@ -278,7 +278,7 @@ def test_simple_self_join(ray_start_regular_shared_2_cpus, left_suffix, right_su
             rsuffix=right_suffix,
         ).reset_index(drop=True)
 
-        pd.testing.assert_frame_equal(expected_pd, joined_pd_sorted)
+        assert rows_same(expected_pd, joined_pd_sorted), "Expected contents to be same"
 
 
 def test_invalid_join_config(ray_start_regular_shared_2_cpus):

--- a/python/ray/data/tests/test_join.py
+++ b/python/ray/data/tests/test_join.py
@@ -263,8 +263,7 @@ def test_simple_self_join(ray_start_regular_shared_2_cpus, left_suffix, right_su
 
         assert 'Field "double" exists 2 times' in str(exc_info.value.cause)
     else:
-
-        joined_pd = pd.DataFrame(joined.take_all()).reset_index(drop=True)
+        joined_pd = joined.to_pandas()
 
         # Join using Pandas (to assert against)
         expected_pd = doubles_pd.join(

